[schema processor] Part 3 - Modifiers and Revisions #12147

MovieStoreGuy opened the pull request:
This is a new PR of https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/12147

No existing changes have happened.

cc: @tigrannajaryan

MovieStoreGuy committed:
[schema processor] Modifier

A modifier allows mutating a signal from a previous version to the
current one and vice versa.
[schema processor] Adding revisions

A revision handles applying the modifiers to incoming signals and
converting a signal to the next version.

tigrannajaryan commented:
@MovieStoreGuy I reviewed https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/12147 earlier. Does this need another round of review?

MovieStoreGuy commented:
> @MovieStoreGuy I reviewed #12147 earlier. Does this need another round of review?

This is the same PR but nether of us could reopen the stale PR.

MovieStoreGuy commented:
I believe I had addressed all the comments left by @tigrannajaryan in the last reviewed pinned.

I know in my delay that schema parser as adopted v1.1, however, I would like to get this out first that works with v1.0 so the semantic convetions can be defined as stable and v1.1 can be added in once this component is added to the release.

bogdandrutu commented on processor/schemaprocessor/internal/translation/modifier_noop.go:
```diff
@@ -0,0 +1,33 @@
+)
+
+// noModify is used when there is no definition present for a given change
+type noModify struct{}
```
nopModifier?

bogdandrutu commented on processor/schemaprocessor/internal/translation/modifier.go:
```diff
@@ -0,0 +1,194 @@
+type Modifier interface {
+	// UpdateAttrs modifies all matching attributes
+	// names with the configured changes.
+	UpdateAttrs(attrs pcommon.Map)
```
Hard to understand the diff between update and revert. After reading only this interface comments I don't know when to use one or the other, this is a bad sign in my opinion.

MovieStoreGuy replied:
I wonder if it is worth while writing it as:

```golang
type VersionIterator interface {
    NextAttributes(...)
    PrevAttributes(...)
}
```

Would that help make it clearer?

bogdandrutu commented on processor/schemaprocessor/internal/translation/modifier.go:
```diff
@@ -0,0 +1,194 @@
+}
+
+// modify is one change that can be made to a signal
+type modify struct {
```
Is this a "rename" action?

bogdandrutu commented on processor/schemaprocessor/internal/translation/modifier.go:
```diff
@@ -0,0 +1,194 @@
+
+func (mod *modify) UpdateAttrs(attrs pcommon.Map) {
+	for from, to := range mod.attrs {
+		if v, match := attrs.Get(from); match {
```
Duplicate code from 78-97 with 104-123

MovieStoreGuy replied:
This will more than likely change once I rework the internals of schema modifications.

bogdandrutu commented on processor/schemaprocessor/internal/translation/modifier.go:
```diff
@@ -0,0 +1,194 @@
+// modifications is used when there is sets of changes
+// that can not be merged together due to dependant changes
+// on signal name
+type modifications []*modify
```
I expect that changes are "applied" in order.

bogdandrutu commented on processor/schemaprocessor/internal/translation/revision.go:
```diff
@@ -0,0 +1,190 @@
+	// sorting revisions easier
+	Version() *Version
+
+	// All is intended to apply to all incoming attributes
+	// that are known as part of this revision.
+	All() Modifier
```
Is this not just a combination of all the others? Where is this used?

MovieStoreGuy replied:
So this is a like for like from the `1.0.0` format, https://opentelemetry.io/docs/reference/specification/schemas/file_format_v1.0.0/#all-section

Basically, if the schemaURL matches then the `all` grouping is applied to every signal.

bogdandrutu commented on processor/schemaprocessor/internal/translation/revision.go:
```diff
@@ -0,0 +1,190 @@
+		mods := make(modifications, l)
+		for i, ch := range def.SpanEvents.Changes {
+			mod := newModify()
+			if ch.RenameAttributes != nil {
```
I think I now understand better the "actions":

So there is a forward change (from v1 -> v2) and a backwards change (v2 ->v1). Not sure how to name the Modifier interface but clearly needs some naming improvements there.

MovieStoreGuy replied:
Yeah, I am still scratching my head on what would be a nice name for the Modifier.

Perhaps something like:
- `SignalValueTranslator` 
- `ValueModifer`
- `RevisionUpdater`

bogdandrutu reviewed (commented):
I will think more to suggest some better design/names.

tigrannajaryan commented on processor/schemaprocessor/internal/translation/modifier.go:
```diff
@@ -0,0 +1,194 @@
+}
+
+func (mod *modify) UpdateAttrs(attrs pcommon.Map) {
+	for from, to := range mod.attrs {
```
I am curious if iterating by `attrs` may result in better performance. Loolkup of `attrs.Get()` likely is slower than `mod.attrs[]` lookup, right, especially if there are a lot of attributes? I am curious how performance of this compares:

```go
attrs.Range(func f(k,v) {
  if v, match := mod.attrs[k]; match {
     ...
  }
})
```

We may also need to verify that the attribute we convert to does not already exist (and error if it does?). 

Also I wonder if we need to make the code work correctly if there are renaming loops (e.g. a->b, b->a). The expectation is that the attributes should be swapped, but the current code will not work correctly in that case. Perhaps we should prohibit that in the spec, currently schemas don't say anything about it.

Finally, since iteration order of maps is not defined, the effects of this function are non-deterministic if there are any renaming loops.

MovieStoreGuy replied:
Potentially, I did forget that `pcommon.Map` is not a conventional golang map so there is a cost there that I would need to come back to.

> We may also need to verify that the attribute we convert to does not already exist (and error if it does?).

That is an interesting question / point, I've not addressed the concern that there is a missing attribute that is defined within the schema. Is there an expected behaviour that is defined within the spec?

> Also I wonder if we need to make the code work correctly if there are renaming loops (e.g. a->b, b->a). The expectation is that the attributes should be swapped, but the current code will not work correctly in that case. Perhaps we should prohibit that in the spec, currently schemas don't say anything about it.

Is this an expected behaviour? I believe I have a test in either this or the next PR that looks to address this in a test where a signal revision has `a --> b` and `b --> a`. Since it only performs one modification per scheme revision. However, I think if you list several modifications within one schema then it could potential happen?
I would need to review this again.

> Finally, since iteration order of maps is not defined, the effects of this function are non-deterministic if there are any renaming loops.

Did you have a inclination of what you'd expect in this scenario? 
My initial thought would be to ignore them and have it so an attribute can only be modified once per revision but I am not fixed to the idea.

tigrannajaryan reviewed (commented):
Thanks for working on this @MovieStoreGuy !

MovieStoreGuy commented on processor/schemaprocessor/internal/translation/revision.go:
```diff
@@ -0,0 +1,190 @@
+
+// Revision represents all changes that are to be
+// applied to a signal at a given version.
+type Revision interface {
```
I had used an interface since at the time the zero value wasn't a safe type.

MovieStoreGuy commented on processor/schemaprocessor/internal/translation/modifier_race_test.go:
```diff
@@ -0,0 +1,76 @@
+		trace := ptrace.NewTraces()
+		traces.CopyTo(trace)
+
+		for i := 0; i < trace.ResourceSpans().Len(); i++ {
+			spans := trace.ResourceSpans().At(i)
+			mod.UpdateAttrs(spans.Resource().Attributes())
+			for t := 0; t < spans.ScopeSpans().Len(); t++ {
+				trace := spans.ScopeSpans().At(t)
+				for s := 0; s < trace.Spans().Len(); s++ {
+					span := trace.Spans().At(s)
+					mod.UpdateAttrs(span.Attributes())
+					mod.UpdateAttrsIf(span.Name(), span.Attributes())
+					mod.UpdateSignal(span)
```
That is right, just trying to check to see if it does what is expected.

MovieStoreGuy committed:
Adding changes

MovieStoreGuy commented:
Still in progress, I just have been unfortunate with trying to find time to resolve this. 

Talking with @mx-psi is keen to help see the implementation finished and take over my current work.

mx-psi commented:
Will discuss further with @MovieStoreGuy and have a look at this next week to see how to continue.

@MovieStoreGuy if you have the time, would you mind going through the conversations and marking as resolved any that no longer need addressing?

tigrannajaryan commented:
@MovieStoreGuy sounds good. Let me know if you need any help from me. I think it is important that we complete this. The lack of the transform processor comes up in our discussions on how to better handle semantic convention changes.

MovieStoreGuy commented on processor/schemaprocessor/internal/translation/revision.go:
```diff
@@ -0,0 +1,190 @@
+	}
+
+	if len(def.Resources.Changes) > 0 {
+		mod := newModify()
+		for _, ch := range def.Resources.Changes {
+			if ch.RenameAttributes == nil {
+				continue
+			}
+			for k, v := range *ch.RenameAttributes {
+				mod.attrs[k] = v
+			}
+		}
```
Yeah, I need to revisit the type and include a constructor for it.

MovieStoreGuy committed:
Adding migrate package.

Following patterns similar to databases where you'd apply changes
and rollback. the Migrate packages follows these designs to help make it
clear what it is trying to do.

MovieStoreGuy commented:
Sorry, I was actively working on this so please excuse any broken updates.

MovieStoreGuy committed:
Removing modiferies to avoid confusion

MovieStoreGuy committed:
Fixing up naming issues

MovieStoreGuy committed:
Adding checks in for building a revision

MovieStoreGuy commented:
cc @tigrannajaryan , @mx-psi ,

I've updated the PR here and using more bare metal types and taken the nomenclature that would be used for database migrations when looking at applying changes to values. From this I've created the `migrate` package that contains this updated pattern and `translation.Revision` is where I am expecting exposed types that will be applied to `pdata`.

Please let me know you're thoughts on this <3 

I am hoping once this is merged, we shouldn't be too far away from having this being merged in an indevelopment state to gathering feedback early.

MovieStoreGuy committed:
Fixing up replaces

tigrannajaryan commented on processor/schemaprocessor/internal/migrate/doc.go:
```diff
@@ -0,0 +1,25 @@
+
+const (
+	_ StateSelctor = iota
+	StateSelctorApply
```
```suggestion
	StateSelectorApply
```

tigrannajaryan commented on processor/schemaprocessor/internal/migrate/attributes.go:
```diff
@@ -0,0 +1,108 @@
+// OrderedAttributeChangeSets allows for `AttributeChangeSet`
+// to be chained together as they are defined within the schema
+// and be applied sequentially to ensure deterministic behavior.
+type AttributeChangeSetSlice []*AttributeChangeSet
```
Nit: this probably can be:
```suggestion
type AttributeChangeSetSlice []AttributeChangeSet
```

MovieStoreGuy replied:
I'd like to keep it as pointers to avoid copying entire values onto the stack for each processing step.

I am conscious that optimising early might be a bad idea without actual information of user feedback.

tigrannajaryan commented on processor/schemaprocessor/internal/migrate/attributes.go:
```diff
@@ -0,0 +1,108 @@
+	rollback map[alias.AttributeKey]alias.AttributeKey
+}
+
+// OrderedAttributeChangeSets allows for `AttributeChangeSet`
```
```suggestion
// AttributeChangeSetSlice allows for `AttributeChangeSet`
```

tigrannajaryan commented on processor/schemaprocessor/internal/migrate/attributes.go:
```diff
@@ -0,0 +1,108 @@
+
+// NewAttributes allows for typed strings to be used as part
+// of the invocation that will be converted into the default string type.
+func NewAttributes[Key AttributeKey, Value AttributeKey](mappings map[Key]Value) *AttributeChangeSet {
```
```suggestion
func NewAttributeChangeSet[Key AttributeKey, Value AttributeKey](mappings map[Key]Value) *AttributeChangeSet {
```

tigrannajaryan replied:
Does this need to be a generic func? It seems like AttributeKey is always going to be a string. Do we ever need to use a different type?

MovieStoreGuy replied:
I've kept it consistent between the other methods since you can't have type constraints without using it as a generic type.

tigrannajaryan commented on processor/schemaprocessor/internal/migrate/attributes.go:
```diff
@@ -0,0 +1,108 @@
+			key, matched = a.rollback[s]
+		}
+		if matched {
+			v.CopyTo(attrs.PutEmpty(key))
```
Can we add a TODO to decide what we should do if the key already exists? Should we overwrite it (as it does now) or we should we stop and return an error? This is probably more of a TODO for schema spec.

tigrannajaryan replied:
The `RemoveIf` contract does not say whether it is ok to modify the attrs from the callback. Can we confirm that this is OK and if so make sure `RemoveIf` comments explicitly bless it?

MovieStoreGuy replied:
Let me follow this one up, because it works and doesn't cause a race condition according to the race detector.

MovieStoreGuy replied:
Alright, apparently it isn't recommend so I have updated the actual work that it is doing.

MovieStoreGuy replied:
> Can we add a TODO to decide what we should do if the key already exists? Should we overwrite it (as it does now) or we should we stop and return an error? This is probably more of a TODO for schema spec.

I've made it so it will prefer the schema rewrite values over the existing values to ensure it is deterministic. Meaning that values that are not part of the schema will be removed.

tigrannajaryan replied:
@mx-psi yes, sounds good to me.

tigrannajaryan commented on processor/schemaprocessor/internal/migrate/attributes_test.go:
```diff
@@ -0,0 +1,280 @@
+				m.PutStr("service.version", "v0.0.1")
+			}),
+		},
+		// Due to the way `pcommon.Map.RemoveIf` works,
+		// it forces both values to be removed from the the attributes.
```
This appears to rely on undocumented behavior of RemoveIf. I think we either need to document it or change our implementation not to rely on undocumented behavior.

tigrannajaryan replied:
Can we add a TODO to fix this?

tigrannajaryan commented on processor/schemaprocessor/internal/migrate/conditional.go:
```diff
@@ -0,0 +1,92 @@
+
+type ConditionalAttributeSetSlice []*ConditionalAttributeSet
+
+func NewConditionalAttributeSet[Key AttributeKey, Value AttributeKey, Match ValueMatch](mappings map[Key]Value, matches ...Match) *ConditionalAttributeSet {
```
Same here, does this need to be generic? It seems only string types are ever needed.

MovieStoreGuy replied:
This one needs to be generic since the schema package defines the [types](go.opentelemetry.io/otel/schema/v1.0/types) package that forces those values to be new types instead of alias for strings.

The generic receiver here means we can take the `can convert to string` values and store them internally as strings with one cast in the constructer. 
Save me having to write optional handlers for each type defined by the `types` package.

tigrannajaryan replied:
I see.

Instead of a string can we use `types.AttributeName`, e.g. `map[types.AttributeName]types.AttributeName`? Shouldn't this work without type casts?

I saw your [other PR ](https://github.com/open-telemetry/opentelemetry-go/pull/4082). I think I would prefer to keep the strong types in the `types` package. If it means we need to use generics here then so be it.

MovieStoreGuy replied:
It looks like I can use `ast.AttributeMap` since that is used every where, but with the conditionals, looks like I would need the generic type for `SpanName`, `EventName`, `MetricName`.

tigrannajaryan commented on processor/schemaprocessor/internal/migrate/conditional.go:
```diff
@@ -0,0 +1,92 @@
+	}
+}
+
+func (ca *ConditionalAttributeSet) Apply(attrs pcommon.Map, values ...string) {
```
Does `values` need to be a slice? It seems we will always call this with one value.

MovieStoreGuy replied:
That is something I needed to follow up on, one thing I was sure of is in the span events section, you can supply a set for applies to `span event names` and `span names`. I wasn't sure if they can be combined to restrict to specific values or if they are treated differently.

tigrannajaryan replied:
Yes, the set to apply to (the condition) can have multiple elements. That is represented by `ca.on` field. The value to check the conditions against is only a single value (the span name or the event name). That is represented by `values` parameter, right? I don't see a situation when we can have multiple elements in `values`. 
Perhaps I am missing something, I may have misunderstood how the Apply() func is supposed to be used.

tigrannajaryan commented on processor/schemaprocessor/internal/migrate/signal.go:
```diff
@@ -0,0 +1,97 @@
+
+// NewSignal will create a `Signal` that will check the provided mappings if it can update a `alias.Signal`
+// and if no values are provided for `matches`, then all values will be updated.
+func NewSignal[Key SignalType, Value SignalType](mappings map[Key]Value) *Signal {
```
Same here: do we have a case where SignalType is not a string?

MovieStoreGuy replied:
Yes, it is defined here `https://pkg.go.dev/go.opentelemetry.io/otel/schema@v0.0.4/v1.0/ast#MetricsChange` as `map[types.MetricName]types.MetricName` which are new types being defined instead of type alias being used.

tigrannajaryan commented on processor/schemaprocessor/internal/migrate/signal.go:
```diff
@@ -0,0 +1,97 @@
+
+// Signal allows for migrating types that
+// implement the `alias.Signal` interface.
+type Signal struct {
```
Perhaps:
```suggestion
type SignalNameChange struct {
```

tigrannajaryan replied:
And maybe we can also rename `alias.Signal` to `alias.NamedSignal`.

tigrannajaryan commented on processor/schemaprocessor/internal/translation/revision_v1.go:
```diff
@@ -0,0 +1,125 @@
+}
+
+func newAttributeChangeSetSliceFromChanges(attrs ast.Attributes) *migrate.AttributeChangeSetSlice {
+	values := make([]*migrate.AttributeChangeSet, 0, 10)
```
Any particular reason to use 10 as capacity?

MovieStoreGuy replied:
It was just a random number I pulled from no where.

tigrannajaryan replied:
Nit: since it has no particular importance I think it can be safely deleted, `values := make([]*migrate.AttributeChangeSet)` should be good enough.

MovieStoreGuy replied:
Unfortunately, (not sure when), but the latest version of go requires at least `len` and an optional `cap`.

Tested it here (to make sure it wasn't a linter getting it in the way)
https://go.dev/play/p/9uNpeSYw6gN

tigrannajaryan replied:
You are right. Don't know what I was thinking. What I probably wanted to say is that you can start with an empty slice since `append` later will happily allocate as many as needed:
```go
var values []*migrate.AttributeChangeSet
```

tigrannajaryan commented on processor/schemaprocessor/internal/translation/testdata/1.6.1:
```diff
@@ -0,0 +1,7 @@
+file_format: 1.0.0
+schema_url: https://opentelemetry.io/schemas/1.6.1
+versions:
+  1.6.1:
+  1.5.0:
+  1.4.0:
+  1.0.0:
```
The file doesn't see to be used anywhere. Same for testdata/1.9.0

MovieStoreGuy replied:
It was previously being used in the original code, but I am waiting to the next version to generate revision slice as part of next PR to help keep PRs reviewable.

tigrannajaryan commented on processor/schemaprocessor/transformer_test.go:
```diff
@@ -30,17 +29,6 @@ import (
 	"go.uber.org/zap/zaptest"
 )
 
-//go:embed testdata/schema.yml
-var schemaContent []byte
-
-func SchemaHandler(t *testing.T) func(wr http.ResponseWriter, r *http.Request) {
-	assert.NotEmpty(t, schemaContent, "SchemaContent MUST not be empty")
-	return func(wr http.ResponseWriter, r *http.Request) {
-		_, err := wr.Write(schemaContent)
-		assert.NoError(t, err, "Must not have issues writing schema content")
-	}
-}
-
```
Is this deleted intentionally? No longer needed?

MovieStoreGuy replied:
I will need to double check, may have accidentally  caught it in git.

MovieStoreGuy replied:
This is intentional , it will be moved to translations in the next PR.

tigrannajaryan reviewed (commented):
I mostly like the change, have just some minor comments/questions.

MovieStoreGuy committed:
Fixing naming

MovieStoreGuy committed:
Fixing up conflict

mx-psi commented on processor/schemaprocessor/internal/migrate/attributes.go:
```diff
@@ -0,0 +1,131 @@
+		if matched {
+			k, updated[key] = key, struct{}{}
+		} else {
+			// TODO: Since the spec hasn't decided the behavior on what
```
Link to an issue for deciding this in the comment?

MovieStoreGuy replied:
Not yet, was raised by @tigrannajaryan in a comment so I will raise it in the morning on the spec repo (or if anyone else gets to it before me).

MovieStoreGuy replied:
Added here https://github.com/open-telemetry/opentelemetry-specification/issues/3497

mx-psi replied:
I think this must be solved before marking the component as alpha and adding it to our official builds, but still I don't think it should block merging this PR in the interest of moving forward with other parts of the implementation

mx-psi commented on processor/schemaprocessor/internal/migrate/doc.go:
```diff
@@ -0,0 +1,25 @@
+// on specific schema fields that are consumed by the translation package.
+package migrate // import "github.com/open-telemetry/opentelemetry-collector-contrib/processor/schemaprocessor/internal/migrate"
+
+type StateSelctor int
```
```suggestion
type StateSelector int
```

(and everywhere else :) )

MovieStoreGuy committed:
Fixing typo

MovieStoreGuy committed:
Updating naming

tigrannajaryan reviewed (approved):
LGTM

MovieStoreGuy committed:
Fixing up ast usage

MovieStoreGuy commented:
Sorry, just removing some of the generics usage since I can use `ast.AttributeMap` where `map[AttributeKey]AttributeKey` was being defined without issue.

mx-psi commented:
Let's wait until EOW to give @bogdandrutu time to review and I will merge if no new concerns are raised :)

MovieStoreGuy committed:
Removing transform files

MovieStoreGuy committed:
FIxing up generated mod files

MovieStoreGuy committed:
Fixing up license headers

mx-psi commented:
@bogdandrutu I am going to merge this, we won't move this out of 'In Development' until we clarify https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/17020#discussion_r1192078334 but I am going to merge this since we have an issue tracking this on the spec. Please open follow up issues if you have any comments
